### PR TITLE
NAS-110514 / 21.08 / UI should show that it failed to retrieve catalogs data

### PR DESCRIPTION
**Pull request by AlexKarpov98**

### CODE REVIEW

**Comment by undsoft** [comment]
What are the steps to test this?

**Comment by denysbutenko** [comment]
How to test it?

**Comment by AlexKarpov98** [comment]
> What are the steps to test this?

This ticket is for unexpected issue on middleware end so no way to test it easily.
I will have to ask Waqar to do it.

**Review by sonicaj — CHANGES_REQUESTED** [review]
We show 2 different error messages if middleware is unable to retrieve catalogs data. We should only show 1 error message which should convey that UI is not able to retrieve catalog data.

Secondly, if we are able to retrieve data for at least 1 catalog, we should not show this error message and show the available applications from the catalog(s) which have `error=False`. Right now that is not true and even if 1 catalog errors out ( is malformed or any other error ), no application is rendered.

https://user-images.githubusercontent.com/17968138/119179026-df061080-ba87-11eb-8550-560251f6dffa.mov

**Comment by AlexKarpov98** [comment]
> We show 2 different error messages if middleware is unable to retrieve catalogs data. We should only show 1 error message which should convey that UI is not able to retrieve catalog data.
> 
> Secondly, if we are able to retrieve data for at least 1 catalog, we should not show this error message and show the available applications from the catalog(s) which have `error=False`. Right now that is not true and even if 1 catalog errors out ( is malformed or any other error ), no application is rendered.
> 
>  Screen.Recording.2021-05-21.at.10.53.46.PM.mov

@sonicaj  Let me make sure.
So there are 2 cases:
1. Fetching catalogs from middleware is unsuccesful (that means we have at least one `error=true` catalog.
2. Fetching catalogs from middleware is successful but empty data.

I think we need to show different 2 error messages respectively. Your thoughts?

File: src/app/helptext/apps/apps.ts [hunk 1]
```diff
@@ -23,10 +23,18 @@ export default {
...
     not_configured: T('Applications not configured'),
     not_running: T('Applications are not running'),
+    no_search_result: T('No Search Results.'),
     no_installed: T('No Applications Installed'),
     no_installed_message: T('Applications you install will automatically appear here. Click below and browse the TrueNAS catalog to get started.'),
   },
 
+  catalogMessage: {
+    loading: T('Loading...'),
+    has_api_error: T('It was unable to retrieve catalogs data.'),
```
[the remark this answers is not in the document]
Comment (AlexKarpov98) [reply src/app/helptext/apps/apps.ts:33]: This is the error message that we show on error popup when middleware is unable to retrieve catalogs.

[the remark this answers is not in the document]
Comment (sonicaj) [reply src/app/helptext/apps/apps.ts]: Yes but can you please clarify what `It` represents in the error message ?

**Comment by sonicaj** [comment]
@AlexKarpov98 the only case i can think of where there will be no apps to show is when official catalog is erroring out only and there are other catalogs which do not error out but there are no apps there ( so quite an unlikely case considering both factors ).

Anyways, in this context what we can be done is that we can show an error message saying that there are no Available Applications. However, currently as we can see in the video 2 error messages are show. This should be consistent and we shouldn't confuse the user by changing error messages. If UI is taking time to load the available applications, then it should be intuitive in the UI that it is loading applications or seeing if there are any available.

**Comment by AlexKarpov98** [comment]
> @AlexKarpov98 the only case i can think of where there will be no apps to show is when official catalog is erroring out only and there are other catalogs which do not error out but there are no apps there ( so quite an unlikely case considering both factors ).
> 
> Anyways, in this context what we can be done is that we can show an error message saying that there are no Available Applications. However, currently as we can see in the video 2 error messages are show. This should be consistent and we shouldn't confuse the user by changing error messages. If UI is taking time to load the available applications, then it should be intuitive in the UI that it is loading applications or seeing if there are any available.


@sonicaj Latest one is implemented in that way. Please check once and let me know.

**Comment by sonicaj** [comment]
@AlexKarpov98 we are unable to build webui with your latest changes.

**Comment by AlexKarpov98** [comment]
> @AlexKarpov98 we are unable to build webui with your latest changes.

Sorry, now it should work.  :)

File: src/app/helptext/apps/apps.ts [hunk 2]
```diff
@@ -23,10 +23,17 @@ export default {
...
     not_running: T('Applications are not running'),
+    no_search_result: T('No Search Results.'),
     no_installed: T('No Applications Installed'),
     no_installed_message: T('Applications you install will automatically appear here. Click below and browse the TrueNAS catalog to get started.'),
   },
 
+  catalogMessage: {
+    loading: T('Loading...'),
+    no_search_result: T('No Search Results.'),
+    no_catalog: T('It was unable to retrieve catalogs data.'),
```
Comment (sonicaj) [inline src/app/helptext/apps/apps.ts:34]: Can you please clarify what `It` is for the user ?

**Review by sonicaj — CHANGES_REQUESTED** [review]

**Review by sonicaj — CHANGES_REQUESTED** [review]
We are also not showing a different error message when UI sees that no catalog has any application available ( provided `error=False` )

File: src/app/helptext/apps/apps.ts (hunk 2 above)
Comment (AlexKarpov98) [reply src/app/helptext/apps/apps.ts:34]: Just copied/pasted from your description. If you have any better one, just let me know.

**Comment by AlexKarpov98** [comment]
> We are also not showing a different error message when UI sees that no catalog has any application available ( provided `error=False` )

So you meant the case when all catalog has `error=False` ?

**Comment by sonicaj** [comment]
> @AlexKarpov98 the only case i can think of where there will be no apps to show is when official catalog is erroring out only and there are other catalogs which do not error out but there are no apps there ( so quite an unlikely case considering both factors ).
> 
> Anyways, in this context what we can be done is that we can show an error message saying that there are no Available Applications. However, currently as we can see in the video 2 error messages are show. This should be consistent and we shouldn't confuse the user by changing error messages. If UI is taking time to load the available applications, then it should be intuitive in the UI that it is loading applications or seeing if there are any available.

@AlexKarpov98 i am referring to this case. So the error message(s) are consistent now but we are not showing "No Applications are Available" message if there are none available provided `error=False` for at least 1 catalog.

File: src/app/helptext/apps/apps.ts (hunk 2 above)
Comment (sonicaj) [reply src/app/helptext/apps/apps.ts:34]: `UI is unable to retrieve Available Application(s)` - @truenas/docs-team can you please advise ?

This is for a case where for some reason we are unable to retrieve any available application ( that reason being erroring out very likely on the catalog(s) end ).

**Comment by AlexKarpov98** [comment]
@sonicaj Fixed error displaying for those 2 cases.

File: src/app/helptext/apps/apps.ts [hunk 3]
```diff
@@ -23,10 +23,18 @@ export default {
...
     not_running: T('Applications are not running'),
+    no_search_result: T('No Search Results.'),
     no_installed: T('No Applications Installed'),
     no_installed_message: T('Applications you install will automatically appear here. Click below and browse the TrueNAS catalog to get started.'),
   },
 
+  catalogMessage: {
+    loading: T('Loading...'),
+    no_search_result: T('No Search Results.'),
+    no_catalog: T('Unable to retrieve catalog data'),
```
Comment (sonicaj) [inline src/app/helptext/apps/apps.ts:34]: `Unable to retrieve Available Application(s)`

File: src/app/helptext/apps/apps.ts [hunk 4]
```diff
@@ -23,10 +23,18 @@ export default {
...
+    no_application: T('No application is available'),
```
Comment (sonicaj) [inline src/app/helptext/apps/apps.ts:35]: `No Application(s) are Available`

**Review by sonicaj — CHANGES_REQUESTED** [review]

**Comment by AlexKarpov98** [comment]
Updated error messages.

**Review by sonicaj — APPROVED** [review]

File: src/app/interfaces/catalog.interface.ts [hunk 5]
```diff
@@ -18,6 +18,7 @@ export interface Catalog {
   trains: {
     [trainName: string]: CatalogTrain;
   };
+  error: boolean;
```
Comment (undsoft) [inline src/app/interfaces/catalog.interface.ts:21]: Is this an API field? If not, please make a separate interface to extend from Catalog.

File: src/app/pages/applications/catalog/catalog.component.html [hunk 6]
```diff
@@ -48,4 +48,5 @@ <h3 textLimiter threshold="15">{{ item.name }}</h3>
       </div>
     </div>
   </mat-card>
+  <entity-empty *ngIf="filteredCatalogApps.length == 0" [conf]="emptyPageConf"></entity-empty>
```
Comment (undsoft) [inline src/app/pages/applications/catalog/catalog.component.html:51]: Here and below == 0 is unnecessary

**Review by undsoft — APPROVED** [review]

File: src/app/interfaces/catalog.interface.ts (hunk 5 above)
Comment (AlexKarpov98) [reply src/app/interfaces/catalog.interface.ts:21]: Yeah, it's an API field.

File: src/app/pages/applications/catalog/catalog.component.html (hunk 6 above)
Comment (AlexKarpov98) [reply src/app/pages/applications/catalog/catalog.component.html:51]: I don't get it, you mean `*ngIf="filteredCatalogApps.length"` ?

Comment (undsoft) [reply src/app/pages/applications/catalog/catalog.component.html:51]: I mean that you can do !filteredCatalogApps.
I guess it's not too important.

Comment (AlexKarpov98) [reply src/app/pages/applications/catalog/catalog.component.html:51]: I don't think it will work here. ![] will give false.

Comment (undsoft) [reply src/app/pages/applications/catalog/catalog.component.html:51]: Sorry, I meant to write `!filteredCatalogApps.length`

**Pull request merged**